subsys/bluetooth/audio/has_client.c

```diff
@@ -0,0 +1,606 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/gatt.h>
+#include <bluetooth/audio/has.h>
+#include <sys/check.h>
+
+#include "has_internal.h"
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_HAS_CLIENT)
+#define LOG_MODULE_NAME bt_has_client
+#include "../common/log.h"
+
+#define HAS_INST(_has) CONTAINER_OF(_has, struct has_inst, has)
+#define HANDLE_IS_VALID(handle) ((handle) != 0x0000)
+
+enum {
+	HAS_DISCOVER_IN_PROGRESS,
+
+	HAS_NUM_FLAGS, /* keep as last */
+};
+
+static struct has_inst {
+	/** Common profile reference object */
+	struct bt_has has;
+
+	/** Profile connection reference */
+	struct bt_conn *conn;
+
+	/** Callbacks registered */
+	const struct bt_has_client_cb *cb;
+
+	/** Internal flags */
+	ATOMIC_DEFINE(flags, HAS_NUM_FLAGS);
+
+	/* GATT procedure parameters */
+	struct {
+		struct bt_uuid_16 uuid;
+		union {
+			struct bt_gatt_read_params read;
+			struct bt_gatt_discover_params discover;
+		};
+	} params;
+
+	struct bt_gatt_subscribe_params features_subscription;
+	struct bt_gatt_subscribe_params control_point_subscription;
+	struct bt_gatt_subscribe_params active_index_subscription;
+} has_insts[CONFIG_BT_MAX_CONN];
+
+static struct has_inst *inst_by_conn(struct bt_conn *conn)
+{
+	struct has_inst *inst = &has_insts[bt_conn_index(conn)];
+
+	if (inst->conn == conn) {
+		return inst;
+	}
+
+	return NULL;
+}
+
+static void inst_cleanup(struct has_inst *inst)
+{
+	bt_conn_unref(inst->conn);
+
+	memset(inst, 0, sizeof(*inst));
+}
+
+static void disconnected(struct bt_conn *conn, uint8_t reason)
+{
+	struct has_inst *inst = inst_by_conn(conn);
+
+	if (inst) {
+		inst_cleanup(inst);
+	}
+}
+
+BT_CONN_CB_DEFINE(conn_cb) = {
+	.disconnected = disconnected,
+};
+
+static enum bt_has_capabilities get_capabilities(const struct has_inst *inst)
+{
+	enum bt_has_capabilities caps = 0;
+
+	if (HANDLE_IS_VALID(inst->control_point_subscription.value_handle)) {
+		caps |= BT_HAS_PRESET_SUPPORT;
+	}
+
+	return caps;
+}
+
+static uint8_t control_point_notify_cb(struct bt_conn *conn,
+				       struct bt_gatt_subscribe_params *params, const void *data,
+				       uint16_t len)
+{
+	/* TODO: Handle Control Point PDU */
+
+	return BT_GATT_ITER_CONTINUE;
+}
+
+static void discover_complete(struct has_inst *inst)
+{
+	BT_DBG("conn %p", (void *)inst->conn);
+
+	atomic_clear_bit(inst->flags, HAS_DISCOVER_IN_PROGRESS);
+
+	inst->cb->discover(inst->conn, 0, &inst->has,
+			   inst->has.features & BT_HAS_FEAT_HEARING_AID_TYPE_MASK,
+			   get_capabilities(inst));
+
+	/* If Active Preset Index supported, notify it's value */
+	if (HANDLE_IS_VALID(inst->active_index_subscription.value_handle)) {
+		inst->cb->active_index(&inst->has, inst->has.active_index);
+	}
+}
+
+static void discover_failed(struct has_inst *inst, int err)
+{
+	BT_DBG("conn %p", (void *)inst->conn);
+
+	inst->cb->discover(inst->conn, err, NULL, 0, 0);
+
+	inst_cleanup(inst);
```

Comment:
Do we need to call `inst_cleanup` before the callback? 

If the application attempts to call e.g. `bt_has_client_discover` in the callback then it will set all the values again, and once it has returned, we will call `inst_cleanup` which discovery may be in progress.

Reply:
Alternatively, as I've also done for some other implementation, we can move the call to `inst_cleanup` to the `bt_has_client_discover` instead, such that we reset before we start discovery instead.

Reply:
Nice catch, I'll go with 1st approach you proposed

Reply:
done